include/oneapi/dpl/internal/reduce_by_segment_impl.h

```diff
@@ -138,6 +167,414 @@ reduce_by_segment_impl(Policy&& policy, InputIterator1 first1, InputIterator1 la
+                                    if (__loc_seg_ends_acc[__j])
+                                    {
+                                        __last_it = true;
+                                        break;
+                                    }
+                                }
+                            }
+                            __agg_collector = __dpl_sycl::__group_broadcast(__item.get_group(), __agg_collector);
+                            __last_it = __dpl_sycl::__group_broadcast(__item.get_group(), __last_it);
+                        }
+
+                        // Check to see if aggregates exist.
+                        // The last group must always stay to write the final index
+                        __ag_exists = __dpl_sycl::__any_of_group(__group, __ag_exists);
+                        if (!__ag_exists && __group_id != __n_groups - 1)
+                            return;
+                    }
+                    // 3b. count the segment ends
+                    for (::std::size_t __i = __start; __i < __end; ++__i)
+                        if (__i == __n - 1 || !__binary_pred(__keys[__i], __keys[__i + 1]))
+                            ++__item_segments;
+
+                    ::std::size_t __prior_segs_in_wg = __dpl_sycl::__exclusive_scan_over_group(
+                        __group, __item_segments, __dpl_sycl::__plus<decltype(__item_segments)>());
+
+                    // 3c. Determine prior index
+                    ::std::size_t __wg_num_prior_segs = __seg_ends_scan_acc[__group_id];
+
+                    // 3d. Second pass over the keys, reidentifying end segments and applying work group
+                    // aggregates if appropriate. Both the key and reduction value are written to the final output at the
+                    // computed index
+                    ::std::size_t __item_offset = 0;
+                    for (::std::size_t __i = __start; __i < __end; ++__i)
+                    {
+                        if (__i == __n - 1 || !__binary_pred(__keys[__i], __keys[__i + 1]))
+                        {
+                            ::std::size_t __idx = __wg_num_prior_segs + __prior_segs_in_wg + __item_offset;
+
+                            // apply the aggregate if it is the first segment end in the workgroup only
+                            if (__prior_segs_in_wg == 0 && __item_offset == 0 && __ag_exists)
+                                __out_values[__idx] = __binary_op(__agg_collector, __out_values[__idx]);
+
+                            ++__item_offset;
+                            // the last item must write the last index's position to return
+                            if (__i == __n - 1)
+                                __end_idx_acc[0] = __idx;
+                        }
+                    }
+                });
+        })
+        .wait();
+
+    return __end_idx.get_host_access()[0] + 1;
```

Comment:
To clarify, should I go ahead and make this function return a future to the end index accessor? If so, I will make this change and the corresponding change to the legacy `__pattern_reduce_by_segment` which is used for types without identities in `algorithm_ranges_impl_hetero.h`

Reply:
After second thought and discussion with @danhoeflinger, would it be better to add this as an issue to address after this is merged? The current mainline implementations of `__pattern_reduce_by_segment` and `*_scan_by_segment` are synchronous. I think it might be cleaner to directly address this in a separate PR especially due to the current size of this PR and time constraints.

Reply:
So, please create a "Reference in new issues"  - from "points" menu, upper right corner.

Reply:
Created this issue.